overlord/restart/restart.go

```diff
@@ -362,3 +362,126 @@ func ReplaceBootID(st *state.State, bootID string) {
+// For tasks that request restarts as a part of their undo, any tasks that previously scheduled
+// restarts as a part of their 'do' will be unscheduled.
+// XXX: will replace FinishTaskWithRestart
+func FinishTaskWithRestart2(t *state.Task, snapName string, status state.Status, restartType RestartType, rebootInfo *boot.RebootInfo) error {
+	switch restartType {
+	case RestartSystem, RestartSystemNow, RestartSystemHaltNow, RestartSystemPoweroffNow:
+		break
+	default:
+		t.SetStatus(status)
+		Request(t.State(), restartType, rebootInfo)
+		return nil
+	}
+
+	chg := t.Change()
+	rt, err := changeRestartContext(chg)
+	if err != nil {
+		return err
+	}
+
+	if snapName == "" {
+		snapName = "snapd"
+	}
+	rt.setParameters(snapName, restartType, rebootInfo)
+
+	// Either invoked with undone or done as the final status. We don't expect
+	// other uses than tasks that end their Doing/Undoing to call this. Let's
+	// only allow these for now as nothing tests with anything else.
+	switch status {
+	case state.DoneStatus, state.UndoneStatus:
+		markTaskForRestart(t, status)
+	default:
+		return fmt.Errorf("unexpected task status when requesting system restart for task: %s", status)
+	}
+
+	chg.Set("restart-ctx", rt)
+	return nil
+}
+
+// PendingForChange checks if a system restart is pending for a change.
+func PendingForChange(st *state.State, chg *state.Change) bool {
+	rm := restartManager(st, "internal error: cannot request a restart before RestartManager initialization")
+	return rm.PendingForSystemRestart(chg)
+}
+
+// TaskWaitForRestart can be used for tasks that need to wait for a pending
+// restart to occur, meant to be used in conjunction with PendingForChange.
+// The task will then be re-run after the restart has occurred.
+// This is only supported for tasks that are currently in Doing/Undoing and is only
+// safe to call from the task itself. After calling this, the task should immediately
+// return.
+func TaskWaitForRestart(t *state.Task) error {
+	// We catch them in Undoing/Doing and restore them to
+	// Do/Undo so they are re-run as we cannot save progress mid-task.
+	switch t.Status() {
+	case state.UndoingStatus:
+		markTaskForRestart(t, state.UndoStatus)
+	case state.DoingStatus:
+		markTaskForRestart(t, state.DoStatus)
+	default:
+		return fmt.Errorf("only tasks currently in progress (doing/undoing) are supported")
+	}
+	return nil
+}
+
+// processRestartForChange must only be called from the change status changed event
+// hook.
+func processRestartForChange(chg *state.Change) error {
+	var rt RestartContext
+	if err := chg.Get("restart-ctx", &rt); err != nil {
+		if errors.Is(err, &state.NoStateError{}) {
+			return fmt.Errorf("change %s is waiting to continue but has not requested any reboots", chg.ID())
+		}
+		return err
+	}
+
+	// clear out the restart context for this change before restarting
+	chg.Set("restart-ctx", nil)
+
+	// perform the restart
+	rt.performReboot(chg.State())
+	return nil
```

Comment:
this code also does not seem tested here

Reply:
fixed